File: app/src/main/java/com/xephorium/crystalnote/ui/base/DrawerActivity.java
package com.xephorium.crystalnote.ui.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.google.android.material.navigation.NavigationView;
import com.xephorium.crystalnote.R;
import com.xephorium.crystalnote.ui.custom.NoteToolbar;
import com.xephorium.crystalnote.ui.util.DisplayUtils;

public class DrawerActivity extends BaseActivity {

    private DrawerLayout drawerLayout;
    private ActionBarDrawerToggle drawerToggle;
    private LinearLayout statusBarMargin;
    private NoteToolbar toolbar;
    private LinearLayout activityContent;
    private NavigationView drawerContent;
    private boolean drawerOpen;
    private boolean drawerAnimating;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.navigation_activity_layout);

        drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        statusBarMargin = (LinearLayout) findViewById(R.id.status_bar_margin);
        toolbar = (NoteToolbar) findViewById(R.id.toolbar);
        activityContent = (LinearLayout) findViewById(R.id.activity_content);
        drawerContent = (NavigationView) findViewById(R.id.drawer_content);
        drawerOpen = false;
        drawerAnimating = false;

        drawerToggle = getActionBarDrawerToggle();
        drawerToggle.setDrawerIndicatorEnabled(false);
        drawerLayout.setDrawerListener(drawerToggle);

        setupStatusBar();
        setupToolbar();
        setupNavDrawer();
    }

    public void setActivityContent(int layoutResource) {
        activityContent.addView(getInflatedView(layoutResource));
    }

    @Override
    public void onBackPressed() {
        if (drawerAnimating)
            return;

        if (drawerOpen) {
            drawerLayout.closeDrawers();
            drawerAnimating = true;
        } else
            super.onBackPressed();
    }

    private void setDrawerContent() {
        // Do Nothing
    }

    public void openDrawer() {
        drawerLayout.openDrawer(GravityCompat.START);
        drawerAnimating = true;
    }

    public void closeDrawer() {
        drawerLayout.closeDrawers();
        drawerAnimating = true;
    }

    public NoteToolbar getToolbar() {
        return toolbar;
    }

    private void setupStatusBar() {
        statusBarMargin.setMinimumHeight(DisplayUtils.getStatusBarHeight(this));
        statusBarMargin.setElevation(getResources().getDimension(R.dimen.toolbar_elevation));
    }

    private void setupToolbar() {
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
    }

    private void setupNavDrawer() {
        drawerContent.getLayoutParams().width = DisplayUtils.getDisplayWidth(this) -
                getResources().getDimensionPixelSize(R.dimen.toolbar_height);
    }

    private View getInflatedView(int layoutResource) {
        LayoutInflater layoutInflater = getLayoutInflater();
        return layoutInflater.inflate(layoutResource, null);
    }

    private ActionBarDrawerToggle getActionBarDrawerToggle() {
        return new ActionBarDrawerToggle(
                this,
                this.drawerLayout,
                R.string.navigation_drawer_open,
                R.string.navigation_drawer_closed) {

            @Override
            public void onDrawerClosed(View view) {
                super.onDrawerClosed(view);
                drawerOpen = false;
                drawerAnimating = false;
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                drawerOpen = true;
                drawerAnimating = false;
            }
        };
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/base/DrawerActivity.kt
package com.xephorium.crystalnote.ui.base

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View

import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.view.GravityCompat

import com.xephorium.crystalnote.R
import com.xephorium.crystalnote.ui.utility.DisplayUtils
import kotlinx.android.synthetic.main.navigation_activity_layout.*

@SuppressLint("Registered")
open class DrawerActivity : BaseActivity() {


    /*--- Variable Declarations ---*/

    private lateinit var drawerToggle: ActionBarDrawerToggle
    private var drawerOpen: Boolean = false
    private var drawerAnimating: Boolean = false


    /*--- Lifecycle Methods ---*/

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.navigation_activity_layout)

        drawerOpen = false
        drawerAnimating = false

        drawerToggle = getActionBarDrawerToggle()
        drawerToggle.isDrawerIndicatorEnabled = false
        drawer_layout.setDrawerListener(drawerToggle)

        setupStatusBar()
        setupToolbar()
        setupNavDrawer()
    }

    override fun onBackPressed() {
        if (drawerAnimating)
            return

        if (drawerOpen) {
            drawer_layout.closeDrawers()
            drawerAnimating = true
        } else
            super.onBackPressed()
    }


    /*--- Public Methods ---*/

    fun setActivityContent(layoutResource: Int) {
        activity_content.addView(getInflatedView(layoutResource))
    }


    private fun setDrawerContent() {
        // Do Nothing
    }

    fun openDrawer() {
        drawer_layout.openDrawer(GravityCompat.START)
        drawerAnimating = true
    }

    fun closeDrawer() {
        drawer_layout.closeDrawers()
        drawerAnimating = true
    }


    /*--- Private Utility Methods ---*/

    private fun setupStatusBar() {
        status_bar_margin.minimumHeight = DisplayUtils.getStatusBarHeight(this)
        status_bar_margin.elevation = resources.getDimension(R.dimen.toolbar_elevation)
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayShowTitleEnabled(false)
    }

    private fun setupNavDrawer() {
        drawer_content.layoutParams.width = DisplayUtils.getDisplayWidth(this) - resources.getDimensionPixelSize(R.dimen.toolbar_height)
    }

    private fun getInflatedView(layoutResource: Int): View {
        val layoutInflater = layoutInflater
        return layoutInflater.inflate(layoutResource, null)
    }

    private fun getActionBarDrawerToggle() = object : ActionBarDrawerToggle(
            this,
            drawer_layout,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_closed) {

        override fun onDrawerClosed(drawerView: View) {
            super.onDrawerClosed(drawerView)
            drawerOpen = false
            drawerAnimating = false
        }

        override fun onDrawerOpened(drawerView: View) {
            super.onDrawerOpened(drawerView)
            drawerOpen = true
            drawerAnimating = false
        }
    }
}
